### PR TITLE
SONAR-12636 Add a disabled state to the toggle component

### PR DESCRIPTION
**Pull request by philippe-perrin-sonarsource**
Add a disabled state to the toggle component, in order to avoid multiple click when calling an api for example.

Sonar-enterprise associated [PR](https://github.com/SonarSource/sonar-enterprise/pull/2243).

### CODE REVIEW

**Review by jeremy-davis-sonarsource — COMMENTED** [review]
Looks all good, but I question the solution itself.
Shouldn't we distinguish between 
- a disabled button (action is not allowed)
- a busy button (action is already in progress)

WDYT?

**Comment by wouter-admiraal-sonarsource** [comment]
You should ping a SC member. Anything that goes into common should be validated by both teams.

**Comment by wouter-admiraal-sonarsource** [comment]
> Looks all good, but I question the solution itself.
> Shouldn't we distinguish between
> 
> * a disabled button (action is not allowed)
> * a busy button (action is already in progress)
> 
> WDYT?

Interesting thought. How would we display a busy button? With a spinner next to it? This might require a wrapper of some sorts (I'd say a `<span>`, inline-flex'd). How would this affect existing layouts? (by being inline, it shouldn't have any negative effects)

**Comment by philippe-perrin-sonarsource** [comment]
> Looks all good, but I question the solution itself.
> Shouldn't we distinguish between
> 
> * a disabled button (action is not allowed)
> * a busy button (action is already in progress)
> 
> WDYT?

To be honest I thought about that but then I realized that I don't think it should be the component responsibility to render a busy state, but rather the consumer responsibility. Plus, what @wouter-admiraal-sonarsource  said. How would it affects current usages of the component ? WDYT ?

**Comment by jeremy-davis-sonarsource** [comment]
Hmm... I was thinking a spinner inside of it. But you're right that the current pattern is to have it next to it. So nevermind, disabled is fine :+1:

File: components/controls/Toggle.tsx [hunk 1]
```diff
@@ -36,7 +37,7 @@ export default class Toggle extends React.PureComponent<Props> {
   };
 
   handleClick = () => {
-    if (this.props.onChange) {
+    if (this.props.onChange && !this.props.disabled) {
```
Comment (siegfried-ehret-sonarsource) [inline components/controls/Toggle.tsx]: @philippe-perrin-sonarsource You should not need to do this.

Comment (siegfried-ehret-sonarsource) [reply components/controls/Toggle.tsx]: Try `<button disabled onClick={() => alert('plop')}>click me!</button>`, onClick is not called.

**Review by siegfried-ehret-sonarsource — APPROVED** [review]

File: components/controls/Toggle.tsx (hunk 1 above)
Comment (philippe-perrin-sonarsource) [reply components/controls/Toggle.tsx]: You're damn right. Good catch.

**Pull request merged**